src/Host/Broker/Impl/Sessions/Session.cs

```diff
@@ -0,0 +1,196 @@
+
+        public Process Process => _process;
+
+        public SessionInfo Info => new SessionInfo {
+            Id = Id,
+            InterpreterId = Interpreter.Info.Id
+        };
+
+        static Session() {
+            using (var stream = new MemoryStream()) {
+                using (var writer = new BinaryWriter(stream, Encoding.UTF8, leaveOpen: true)) {
+                    writer.Write(ulong.MaxValue - 1);
+                    writer.Write(0UL);
+                    writer.Write("!End".ToCharArray());
+                    writer.Write((byte)0);
+                    writer.Write("[]".ToCharArray());
+                    writer.Write((byte)0);
+                }
+
+                _endMessage = stream.ToArray();
+            }
+        }
+
+        internal Session(SessionManager manager, string id, Interpreter interpreter, IIdentity user) {
+            Manager = manager;
+            Id = id;
+            Interpreter = interpreter;
+            User = user;
+        }
+
+        public void StartHost(SecureString password, ILogger outputLogger, ILogger messageLogger) {
+            string brokerPath = Path.GetDirectoryName(typeof(Program).Assembly.GetAssemblyPath());
+            string rhostExePath = Path.Combine(brokerPath, RHostExe);
+
+            var psi = new ProcessStartInfo(rhostExePath) {
+                UseShellExecute = false,
+                CreateNoWindow = false,
+                Arguments = $"--rhost-name {Id}",
+                RedirectStandardInput = true,
+                RedirectStandardOutput = true,
+                RedirectStandardError = true
+            };
+
+            var shortHome = new StringBuilder(NativeMethods.MAX_PATH);
+            NativeMethods.GetShortPathName(Interpreter.Info.Path, shortHome, shortHome.Capacity);
+            psi.EnvironmentVariables["R_HOME"] = shortHome.ToString();
+            psi.EnvironmentVariables["PATH"] = Interpreter.Info.BinPath + ";" + Environment.GetEnvironmentVariable("PATH");
+
+            if (password != null) {
+                psi.UserName = User.Name;
+                psi.Password = password;
+            }
+
+            _process = new Process {
+                StartInfo = psi,
+                EnableRaisingEvents = true,
+            };
+
+            _process.ErrorDataReceived += (sender, e) => {
+                if (outputLogger != null) {
+                    outputLogger.LogTrace(e.Data);
+                }
+            };
+
+            _process.Exited += delegate {
+                _pipe = null;
+            };
+
+            _process.Start();
+            _process.BeginErrorReadLine();
+
+            _pipe = new MessagePipe(messageLogger);
+            var hostEnd = _pipe.ConnectHost();
+
+            ClientToHostWorker(_process.StandardInput.BaseStream, hostEnd).DoNotWait();
+            HostToClientWorker(_process.StandardOutput.BaseStream, hostEnd).DoNotWait();
+        }
+
+        //public void StartHost() {
```

Comment:
Alternative code that used `ProcessHelpers`. Didn't get it to fully work yet, but this will be the basis for future rework here.